Fix quoting for omprog, improg, mmexternal

paulfertser opened the pull request:
This changes the current behaviour from honouring the double quotes in any part of the argument but leaving them in place and passing to the executed binary to requiring quotes exactly at the beginning and at the end of a multi-word argument, and not including them in the actual call.

Testcases added to ensure the expected results.

Fixes #4249.

I will send the corresponding change to rsyslog-doc repository later if you give the green light to this.

paulfertser commented:
@rgerhards Hey :) I really tried to understand the CI failures but they seem to be non-deterministic. Here we get one error, on this test github pull request with exactly the same code there was another error https://github.com/paulfertser/rsyslog/actions/runs/5088686392/jobs/9145401391 and it's anyway not related to the part of the code I'm changing.
I know you're afk for the coming days, would appreciate your feedback when you return, have a nice weekend.

paulfertser commented:
Gentle reminder about this one. @rgerhards , do you want me to rebase on current HEAD or something else needed to proceed?

paulfertser commented:
Hello @rgerhards ,
Can you please provide some feedback on this one? I tried hard to make proper tests and provide rationale and explanations, is anything still missing?

rgerhards commented:
sry, I overlooked this last may. I'll have a look...

rgerhards commented:
> This changes the current behaviour from honouring the double quotes in any part of the argument but leaving them in place

As I general comment, the rsyslog project does not like breaking changes. This one looks especially problematic. It would be good if you could give some hard reasoning, why this is actually needed and why there are no other ways to reach that functionality.

Even then, we would need to guard it be a config setting, because otherwise we would potentially break a lot of configurations. Rsyslog users expect that no matter how new a version, their existing config continues to work (the good ol'e datacenter paradigm when we had no fear of new releases...).

Re CI, yes it unfortunately is flaky. Timing is critical and CI machines are sometimes overwhelemed (even with other processes on the same bare metal). We could synchronize all and everything, but then the tested and executed code would not have much in common :-(

rgerhards commented:
To me, it looks like the PR re-implement this example from the omprog doc page:

```
module(load="omprog")

action(type="omprog"
       binary="/path/to/log.sh p1 p2 --param3=\"value 3\""
       template="RSYSLOG_TraditionalFileFormat")
```

Note the space in param3.

paulfertser commented:
Hello,

Thank you for taking a look!

> > This changes the current behaviour from honouring the double quotes in any part of the argument but leaving them in place
> 
> As I general comment, the rsyslog project does not like breaking changes. This one looks especially problematic. It would be good if you could give some hard reasoning, why this is actually needed and why there are no other ways to reach that functionality.

That's understandable. Let me try to explain why I thought it's an "ok" idea to offer. First of all, it fixes real bug report affecting several people, mentioned in the commit message. It doesn't seem to be possible to fix it in any other way short of providing an alternative option. Second, the current way to treat quotes in multi-word arguments is breaking assumptions, people are used to how POSIX shells treat quotes and that the commands get the arguments with the quotes removed, so I consider the current behaviour to be plain buggy, it doesn't allow calling external programs properly without additional wrappers, shell evals etc. I wonder how many existing configurations can actually be counting on such an unusual behaviour.

As an example of why it is needed, https://github.com/rsyslog/rsyslog-doc/pull/1007 I think is a good one, as it shows how a very generic and simple (thus correct and unsurprising) script applicable to all other rsyslog omprog usecases can be used without additional helpers.

> Even then, we would need to guard it be a config setting, because otherwise we would potentially break a lot of configurations. Rsyslog users expect that no matter how new a version, their existing config continues to work (the good ol'e datacenter paradigm when we had no fear of new releases...).

If it's absolutely required, OK, I can add another option to omprog action, probably `newQuoting="on"` but that would add to the bloat, retaining buggy code and interfaces for backwards compatibility is so microsoft-ish... I'm open to other ideas of course, it would be nice if you could take another fresh look at the original bug report and my omprog-deque msmtp example, as I'm naturally already too biased towards the solution I was working on.
 
> Re CI, yes it unfortunately is flaky. Timing is critical and CI machines are sometimes overwhelemed (even with other processes on the same bare metal). We could synchronize all and everything, but then the tested and executed code would not have much in common :-(

I do not mind flaky CI, it's just not outright obvious whether you expect contributors to worry about the failures reported on "github pull requests".

rgerhards commented:
I think I see your point, but I need to probably do a number of tests to confirm there really is a problem (we need to be crystal-clear here, especially when we declare it a bug and try to fix with a breaking change). Unfortunately, I am busy this week and out of office next. If you don't hear back by mid-may, pls ping me. I have assigned a milestone, so the issue should at least not get out of "vision".

paulfertser commented:
It is possible to solve the problem by keeping the existing syntax and filtering out the quotes but that would 1. still not allow embedding a literal double quote in the argument when really desired; 2. require creating a new string and filling it char-by-char rather than just calling `es_newStrFromSubStr`. It might be a reasonable trade-off between backwards compatibility and bug fixing after all. It also would still have the potential to break existing configs if someone's really counting on the quotes to be there in the passed arguments. Yet another option is to keep the code as it is and provide additional shell wrapper which would cover all the usecases somehow (probably other than embedding double quote as they would all be stripped one way or another). Or just document the limitations and clarify that if the current option handling is not suitable one must embed all the options in external script.
So yes, plenty of things to consider here, I hope at least the problem is clear enough now to make an informed decision once you have time for it.

rgerhards commented:
I would really like to dig a bit deeper, as I begin to think you are right its a bug. But it's several years since this code was written and I want to find out what were the reasons behind that.

rgerhards commented:
OK, sry for my ignorance. I had a strange misunderstanding. There indeed is a bug. Somehow I thought (even after reading multiple times) it was meant the the quotes were removed and were intended to stay...

Maybe because this is the question so often when talking about "such" templates. Anyhow, now it's clear.

rgerhards commented:
As this really is a bug that affects every well-behaving problem, I think we do not keep backwards-compatibility. Especially as I think it is hard to envision that someone actually depends on the quotes. I would tend to say that always a "starter bash script" was used in cases where it mattered.

rgerhards commented:
The CI issue is a flake, so it is fine to merge - will happen soon.

rgerhards commented:
Thanks again for the PR - and your patience and persistence! Much appreciated.

paulfertser commented:
I'm happy to help! Hope you won't regret the change :)
The idea to require quotes for multi word arguments exactly at the beginning and the end is a bit controversial but that's the best I could come up with without adding escaping for quotes or spaces and without complicating the code too much. At least now the arguments handling is covered by the test cases.
I'm available to discuss and tweak the corresponding change to rsyslog-doc https://github.com/rsyslog/rsyslog-doc/pull/1007 if desired.

rgerhards commented:
> I'm happy to help! Hope you won't regret the change :)

> The idea to require quotes for multi word arguments exactly at the beginning and the end is a bit controversial but that's the best I could come up with without adding escaping for quotes or spaces and without complicating the code too much.

I agree to this approach. I thought about it, but I did not come up with a valid case, where at the end is not an unescaped quote. In CLI, that would mean bash expects more input. If it poses a problem (I really doubt it), we may later add an error message to those cases.

FYI my ChangeLog entry for this change:
```
- 2024-04-18: [o|i]mprog/mmexternal bugfix: invalid command line parameter passing
  Quoted command line parameters were incorrectly passed to the called program. This
  resulted in unexpected behaviour. For example, in
  bash -c "echo test"
  bash got passed parameters '-c" and '"echo test"'. Note the double quotes in
  '"echo test"'. These needed to be removed by Unix/Linux standards but were instead
  passed to bash. That in turn resulted in invalid execution. With the fix, the
  behaviour is now correct. Now, bash gets passed '-c' and 'echo test'.
  This actually is a BREAKING CHANGE. However, it was outright wrong behaviour from
  the beginning. We assume that people either never noticed it (because they did not
  use quoted parameters) or used some workaroud, likely a "starter script", which
  performed the right task. As such, we do expect that only a very limited set of
  installations might be affected by the change.
  Even more important, we would need to change the default behaviour in any case,
  because the previous handling was obviosuly unacceptable. As such, there was no
  way to keep rsyslog perform the previous action.
  Thanks to Paul Fertser for the patch.
```

> At least now the arguments handling is covered by the test cases.

Well, I would tend to say that there were tests, but they were based on the same wrong assumption made during development.

> I'm available to discuss and tweak the corresponding change to rsyslog-doc [rsyslog/rsyslog-doc#1007](https://github.com/rsyslog/rsyslog-doc/pull/1007) if desired.

Looks good to me, but if you could rebase, things would be smoother. We had some CI issues at the time you submitted it, these are probably now fixed (still a lot of moving parts at the moment because we optimize the doc).

paulfertser commented:
> FYI my ChangeLog entry for this change:

Looks convincing to me, thank you.

> > At least now the arguments handling is covered by the test cases.
> 
> Well, I would tend to say that there were tests, but they were based on the same wrong assumption made during development.

Yes, I see, parameters were tested but not how quoting and splitting worked, `$@` just printed them all: https://github.com/rsyslog/rsyslog/commit/2847fa3aa0974c6b8400279fa176d47df7968865#diff-a9d0794f19daf54078e936787ea45fd6849bc96e61579f588d220c745dbae696R6

> > I'm available to discuss and tweak the corresponding change to rsyslog-doc [rsyslog/rsyslog-doc#1007](https://github.com/rsyslog/rsyslog-doc/pull/1007) if desired.
> 
> Looks good to me, but if you could rebase, things would be smoother. We had some CI issues at the time you submitted it, these are probably now fixed (still a lot of moving parts at the moment because we optimize the doc).

Sure, rebased and pushed. However CI is still problematic (guess two seconds wait isn't always enough), it built the first version of the patch even though it was after me triggering a new build by a force-push. I also noticed the URL for the generated artifact has a redirect to a server which seems to drop (rather than reject) TCP connections to HTTPS port so my browser "hanged" after automatically trying to force https for that (using an add-on).
